File: python/lsst/afw/geom/skyWcs.py
# This file is part of afw.
#
# Developed for the LSST Data Management System.
# This product includes software developed by the LSST Project
# (https://www.lsst.org).
# See the COPYRIGHT file at the top-level directory of this distribution
# for details of code ownership.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import numpy as np
import scipy

from lsst.utils import continueClass
import lsst.geom
from ._python import reduceTransform
from ._geom import (SkyWcs, makeCdMatrix, makeFlippedWcs, makeModifiedWcs,
                    makeSkyWcs, makeTanSipWcs, makeWcsPairTransform,
                    getIntermediateWorldCoordsToSky, getPixelToIntermediateWorldCoords)
from ._hpxUtils import makeHpxWcs

__all__ = ["SkyWcs", "makeCdMatrix", "makeFlippedWcs", "makeSkyWcs",
           "makeModifiedWcs", "makeTanSipWcs", "makeWcsPairTransform",
           "getIntermediateWorldCoordsToSky", "getPixelToIntermediateWorldCoords",
           "makeHpxWcs"]


@continueClass
class SkyWcs:  # noqa: F811
    def pixelToSkyArray(self, x, y, degrees=False):
        """
        Convert numpy array pixels (x, y) to numpy array sky (ra, dec)
        positions.

        Parameters
        ----------
        x : `np.ndarray`
            Array of x values.
        y : `np.ndarray`
            Array of y values.
        degrees : `bool`, optional
            Return ra, dec arrays in degrees if True.

        Returns
        -------
        ra : `np.ndarray`
            Array of Right Ascension.  Units are radians unless
            degrees=True.
        dec : `np.ndarray`
            Array of Declination.  Units are radians unless
            degrees=True.
        """
        xy = np.vstack((x, y))
        ra, dec = np.vsplit(self.getTransform().getMapping().applyForward(xy), 2)
        ra %= (2.*np.pi)

        if degrees:
            return np.rad2deg(ra.ravel()), np.rad2deg(dec.ravel())
        else:
            return ra.ravel(), dec.ravel()

    def skyToPixelArray(self, ra, dec, degrees=False):
        """
        Convert numpy array sky (ra, dec) positions to numpy array
        pixels (x, y).

        Parameters
        ----------
        ra : `np.ndarray`
            Array of Right Ascension.  Units are radians unless
            degrees=True.
        dec : `np.ndarray`
            Array of Declination.  Units are radians unless
            degrees=True.
        degrees : `bool`, optional
            Input ra, dec arrays are degrees if True.

        Returns
        -------
        x : `np.ndarray`
            Array of x values.
        y : `np.ndarray`
            Array of y values.
        """
        radec = np.vstack((ra, dec))
        if degrees:
            radec = np.deg2rad(radec)

        x, y = np.vsplit(self.getTransform().getMapping().applyInverse(radec), 2)

        return x.ravel(), y.ravel()

    def getRelativeRotationToWcs(self, otherWcs):
        """Get the difference in sky rotation angle to the specified wcs.

        Ignoring location on the sky, if another wcs were atop this one,
        what would the difference in rotation be? i.e. for

        otherWcs = createInitialSkyWcsFromBoresight(radec, rotation, detector)

        what is the value that needs to be added to ``self.rotation`` (or
        subtracted from `other.rotation``) to align them?

        Parameters
        ----------
        otherWcs : `lsst.afw.geom.SkyWcs`
            The wcs to calculate the angle to.

        Returns
        -------
        angle : `lsst.geom.Angle`
            The angle between this and the supplied wcs,
            over the half-open range [0, 2pi).
        """
        # Note: tests for this function live in
        # obs_lsst/tests/test_afwWcsUtil.py due to the need for an easy
        # constructor and instantiated detector, and the fact that afw
        # cannot depend on obs_base or obs_lsst.

        m1 = self.getCdMatrix()
        m2 = otherWcs.getCdMatrix()

        svd1 = scipy.linalg.svd(m1)
        svd2 = scipy.linalg.svd(m2)

        m1rot = np.matmul(svd1[0], svd1[2])
        m2rot = np.matmul(svd2[0], svd2[2])

        v_rot = [1, 0]

        v_rot = np.matmul(v_rot, m1rot)  # rotate by wcs1
        v_rot = np.matmul(v_rot, m2rot.T)  # rotate _back_ by wcs2

        rotation = np.arctan2(v_rot[1], v_rot[0])
        rotation = rotation % (2*np.pi)
        return lsst.geom.Angle(rotation, lsst.geom.radians)


SkyWcs.__reduce__ = reduceTransform

File: python/lsst/afw/geom/_skyWcs.cc
/*
 * This file is part of afw.
 *
 * Developed for the LSST Data Management System.
 * This product includes software developed by the LSST Project
 * (https://www.lsst.org).
 * See the COPYRIGHT file at the top-level directory of this distribution
 * for details of code ownership.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "nanobind/nanobind.h"
#include <lsst/cpputils/python.h>
#include "nanobind/eigen/dense.h"

#include <memory>

#include "astshim.h"
#include "Eigen/Core"
#include "nanobind/stl/vector.h"
#include "ndarray/nanobind.h"

#include "lsst/geom.h"
#include "lsst/afw/geom/Transform.h"
#include "lsst/afw/typehandling/Storable.h"
#include "lsst/afw/geom/SkyWcs.h"
#include "lsst/afw/table/io/python.h"  // for addPersistableMethods

namespace nb = nanobind;
using namespace nb::literals;

namespace lsst {
namespace afw {
namespace geom {
namespace {

void declareSkyWcs(lsst::cpputils::python::WrapperCollection &wrappers) {
    wrappers.wrap([](auto &mod) {
        mod.def("makeCdMatrix", makeCdMatrix, "scale"_a, "orientation"_a = 0 * lsst::geom::degrees,
                "flipX"_a = false);
        mod.def("makeFlippedWcs", makeFlippedWcs, "wcs"_a, "flipLR"_a, "flipTB"_a, "center"_a);
        mod.def("makeModifiedWcs", makeModifiedWcs, "pixelTransform"_a, "wcs"_a, "modifyActualPixels"_a);
        mod.def("makeSkyWcs",
                (std::shared_ptr<SkyWcs>(*)(lsst::geom::Point2D const &, lsst::geom::SpherePoint const &,
                                            Eigen::Matrix2d const &, std::string const &))makeSkyWcs,
                "crpix"_a, "crval"_a, "cdMatrix"_a, "projection"_a = "TAN");
        mod.def("makeSkyWcs", (std::shared_ptr<SkyWcs>(*)(daf::base::PropertySet &, bool))makeSkyWcs,
                "metadata"_a, "strip"_a = false);
        mod.def("makeSkyWcs",
                (std::shared_ptr<SkyWcs>(*)(TransformPoint2ToPoint2 const &, lsst::geom::Angle const &, bool,
                                            lsst::geom::SpherePoint const &, std::string const &))makeSkyWcs,
                "pixelsToFieldAngle"_a, "orientation"_a, "flipX"_a, "boresight"_a, "projection"_a = "TAN");
        mod.def("makeTanSipWcs",
                (std::shared_ptr<SkyWcs>(*)(lsst::geom::Point2D const &, lsst::geom::SpherePoint const &,
                                            Eigen::Matrix2d const &, Eigen::MatrixXd const &,
                                            Eigen::MatrixXd const &))makeTanSipWcs,
                "crpix"_a, "crval"_a, "cdMatrix"_a, "sipA"_a, "sipB"_a);
        mod.def("makeTanSipWcs",
                (std::shared_ptr<SkyWcs>(*)(lsst::geom::Point2D const &, lsst::geom::SpherePoint const &,
                                            Eigen::Matrix2d const &, Eigen::MatrixXd const &,
                                            Eigen::MatrixXd const &, Eigen::MatrixXd const &,
                                            Eigen::MatrixXd const &))makeTanSipWcs,
                "crpix"_a, "crval"_a, "cdMatrix"_a, "sipA"_a, "sipB"_a, "sipAp"_a, "sipBp"_a);
        mod.def("makeWcsPairTransform", makeWcsPairTransform, "src"_a, "dst"_a);
        mod.def("getIntermediateWorldCoordsToSky", getIntermediateWorldCoordsToSky, "wcs"_a,
                "simplify"_a = true);
        mod.def("getPixelToIntermediateWorldCoords", getPixelToIntermediateWorldCoords, "wcs"_a,
                "simplify"_a = true);
    });
    wrappers.wrapType(
            nb::class_<SkyWcs, typehandling::Storable>(wrappers.module, "SkyWcs"),
            [](auto &mod, auto &cls) {
                cls.def(nb::init<daf::base::PropertySet &, bool>(), "metadata"_a, "strip"_a = false);
                cls.def(nb::init<ast::FrameDict const &>(), "frameDict"_a);

                cls.def("__eq__", &SkyWcs::operator==, nb::is_operator());
                cls.def("__ne__", &SkyWcs::operator!=, nb::is_operator());

                table::io::python::addPersistableMethods<SkyWcs>(cls);

                cls.def("copyAtShiftedPixelOrigin", &SkyWcs::copyAtShiftedPixelOrigin, "shift"_a);
                cls.def("getFitsMetadata", &SkyWcs::getFitsMetadata, "precise"_a = false);
                cls.def("getPixelScale",
                        (lsst::geom::Angle(SkyWcs::*)(lsst::geom::Point2D const &) const) &
                                SkyWcs::getPixelScale,
                        "pixel"_a);
                cls.def("getPixelScale", (lsst::geom::Angle(SkyWcs::*)() const) & SkyWcs::getPixelScale);
                cls.def("getPixelOrigin", &SkyWcs::getPixelOrigin);
                cls.def("getSkyOrigin", &SkyWcs::getSkyOrigin);
                cls.def("getCdMatrix",
                        (Eigen::Matrix2d(SkyWcs::*)(lsst::geom::Point2D const &) const) & SkyWcs::getCdMatrix,
                        "pixel"_a);
                cls.def("getCdMatrix", (Eigen::Matrix2d(SkyWcs::*)() const) & SkyWcs::getCdMatrix);
                cls.def("getTanWcs", &SkyWcs::getTanWcs, "pixel"_a);
                cls.def("getFrameDict", [](SkyWcs const &self) { return self.getFrameDict()->copy(); });
                cls.def("getTransform", &SkyWcs::getTransform);

                cls.def_prop_ro("isFits", &SkyWcs::isFits);
                cls.def_prop_ro("isFlipped", &SkyWcs::isFlipped);
                cls.def("linearizePixelToSky",
                        (lsst::geom::AffineTransform(SkyWcs::*)(lsst::geom::SpherePoint const &,
                                                                lsst::geom::AngleUnit const &) const) &
                                SkyWcs::linearizePixelToSky,
                        "coord"_a, "skyUnit"_a);
                cls.def("linearizePixelToSky",
                        (lsst::geom::AffineTransform(SkyWcs::*)(lsst::geom::Point2D const &,
                                                                lsst::geom::AngleUnit const &) const) &
                                SkyWcs::linearizePixelToSky,
                        "coord"_a, "skyUnit"_a);
                cls.def("linearizeSkyToPixel",
                        (lsst::geom::AffineTransform(SkyWcs::*)(lsst::geom::SpherePoint const &,
                                                                lsst::geom::AngleUnit const &) const) &
                                SkyWcs::linearizeSkyToPixel,
                        "coord"_a, "skyUnit"_a);
                cls.def("linearizeSkyToPixel",
                        (lsst::geom::AffineTransform(SkyWcs::*)(lsst::geom::Point2D const &,
                                                                lsst::geom::AngleUnit const &) const) &
                                SkyWcs::linearizeSkyToPixel,
                        "coord"_a, "skyUnit"_a);
                cls.def("pixelToSky",
                        (lsst::geom::SpherePoint(SkyWcs::*)(lsst::geom::Point2D const &) const) &
                                SkyWcs::pixelToSky,
                        "pixel"_a);
                cls.def("pixelToSky",
                        (lsst::geom::SpherePoint(SkyWcs::*)(double, double) const) & SkyWcs::pixelToSky,
                        "x"_a, "y"_a);
                cls.def("pixelToSky",
                        (std::vector<lsst::geom::SpherePoint>(SkyWcs::*)(
                                std::vector<lsst::geom::Point2D> const &) const) &
                                SkyWcs::pixelToSky,
                        "pixel"_a);
                cls.def("skyToPixel",
                        (lsst::geom::Point2D(SkyWcs::*)(lsst::geom::SpherePoint const &) const) &
                                SkyWcs::skyToPixel,
                        "sky"_a);
                cls.def("skyToPixel",
                        (std::vector<lsst::geom::Point2D>(SkyWcs::*)(
                                std::vector<lsst::geom::SpherePoint> const &) const) &
                                SkyWcs::skyToPixel,
                        "sky"_a);
                // Do not wrap getShortClassName because it returns the name of the class;
                // use `<class>.__name__` or `type(<instance>).__name__` instead.
                // Do not wrap readStream or writeStream because C++ streams are not easy to wrap.
                cls.def_static("readString", &SkyWcs::readString);
                cls.def("writeString", &SkyWcs::writeString);

                cpputils::python::addOutputOp(cls, "__str__");
                // For repr, we could instead call writeString for the very long AST Frame/Mapping output.
                cpputils::python::addOutputOp(cls, "__repr__");
            });
}
}  // namespace
void wrapSkyWcs(lsst::cpputils::python::WrapperCollection &wrappers) {
    wrappers.addInheritanceDependency("lsst.afw.table.io");
    wrappers.addInheritanceDependency("lsst.afw.typehandling");
    wrappers.addSignatureDependency("astshim");
    declareSkyWcs(wrappers);
}
}  // namespace geom
}  // namespace afw
}  // namespace lsst
